### PR TITLE
[query] Add subset method to SBaseStructCode

### PR DESCRIPTION
**Pull request by chrisvittal**
The subset method is used to construct SSubsetStructCode

### CODE REVIEW

**Comment by chrisvittal** [comment]
This is completely unused, but will need to be used in the ordering changes.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala [hunk 1]
```diff
@@ -23,9 +24,15 @@ trait SBaseStructValue extends SValue {
   def loadField(cb: EmitCodeBuilder, fieldName: String): IEmitSCode = loadField(cb, pt.fieldIdx(fieldName))
 }
 
-trait SBaseStructCode extends SCode {
+trait SBaseStructCode extends SCode { self =>
+  def st: SStruct
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:28]: this isn't valid -- these can represent tuples too.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala [hunk 2]
```diff
@@ -23,9 +24,15 @@ trait SBaseStructValue extends SValue {
...
+
   def memoize(cb: EmitCodeBuilder, name: String): SBaseStructValue
 
   def memoizeField(cb: EmitCodeBuilder, name: String): SBaseStructValue
-}
 
+  def subset(fieldNames: String*): SSubsetStructCode = {
+    val st = SSubsetStruct(self.st, fieldNames.toIndexedSeq)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:35]: match on `st` and fail if it's a tuple, maybe?

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala [hunk 3]
```diff
@@ -23,9 +24,15 @@ trait SBaseStructValue extends SValue {
...
+    new SSubsetStructCode(st, self.asPCode.asBaseStruct) // FIXME, should be sufficient to just use self here
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:36]: yep, OK for now.

**Review by tpoterba — CHANGES_REQUESTED** [review]
Nice! minor comment about struct/tuple clashes.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala (hunk 1 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:28]: There's no issue here. There is no STuple, the code need only be a base struct.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala (hunk 2 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:35]: There is no STuple.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:35]: right, I mean match on the virtual type

Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:35]: Oops, haven't PRed that yet. You can add a `def virtualType: Type = pType.virtualType` to the `sType` trait.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:28]: ah, right. Carry on.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala (hunk 2 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:35]: I don't see the purpose of such a check.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SStruct.scala:35]: OK, I think you're right that we don't need this right now. I'll keep an eye out for usages that might motivate this check though.

**Review by tpoterba — APPROVED** [review]

**Pull request merged**